libraries/chain/controller.cpp

```diff
@@ -1212,23 +1218,28 @@ struct controller_impl {
                      ("lib_num", lib_num)("bn", fork_db_root_block_num()) );
       }
 
-      const uint32_t if_lib = if_irreversible_block_num;
+      uint32_t if_lib = block_header::num_from_id(if_irreversible_block_id);
       const uint32_t new_lib = if_lib > 0 ? if_lib : fork_db_head_irreversible_blocknum();
 
       if( new_lib <= lib_num )
          return;
 
       auto mark_branch_irreversible = [&, this](auto& forkdb) {
-         auto branch = forkdb.fetch_branch( fork_db_head(forkdb, irreversible_mode())->id(), new_lib );
+         auto branch = (if_lib > 0) ? forkdb.fetch_branch( if_irreversible_block_id, new_lib)
+                                    : forkdb.fetch_branch( fork_db_head(forkdb, irreversible_mode())->id(), new_lib );
          try {
+            auto should_process = [&](auto& bsp) {
+               return read_mode == db_read_mode::IRREVERSIBLE || bsp->is_valid();
+            };
```

Comment:
Why is this necessary? I thought that lib and its ancestors are guaranteed to be valid.

Reply:
No, they are not guaranteed to be valid. We are putting non-validated blocks into the fork database. A different branch than our best head can become irreversible which means we have not necessarily validated them. The `lib_advance_test.py` actually triggered this case.

An alternative more complex implementation would be to do a fork switch so we can move irreversible all the way at this point. Instead this moves irreversible as much as possible right here and then waits for a `maybe_switch_forks` call to happen *soon* to switch to the new branch and then move irreversible after that. *soon* in this case should be before the next produced block or on the next received block.

I'll add a comment that explains this.

Reply:
Even if we have not validated the block, it became irreversible because we received a strong QC making it irreversible, so I am not sure what benefit there is to not accept it as irreversible. Even if it did not validate, I don't know what we can do since a quorum of finalizers already accepted it.

Reply:
It is our job as a validating node to validate blocks. If some disaster scenario happens we want to be in a good state. Also our block log should not contain data that is not in our state. We need a consistent state and block log in case we have to shutdown. Not even sure what kind of strange corner cases there would be for a node that has blocks in the block log that are not incorporated into the state.

Reply:
Yes, it would be split. But who is to say this isn't the correct chain. Manual intervention is needed.

Reply:
Do you agree that the node might as well terminate in that case?

Reply:
@heifner I modified your comment by mistake. Sorry!

Reply:
> But who is to say this isn't the correct chain.

A quorum of finalizers said so.

Reply:
> @heifner I modified your comment by mistake. Sorry!

Done that many times myself. No worries.

Reply:
OK, I guess I understand the logic here. We expect that, even if the block made irreversible by the received QC has not been validated yet, we will validate it eventually and then be able to commit it to the state. Seems like a weird state to be in though.